From 0d2e955387d53ed3de4f6f06a3fe8e9403624e3d Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 1 Mar 2022 14:35:47 +0100
Subject: [PATCH 031/954] fix(st-ddr): manage quasi-dynamic register
 modifications

Manage the quasi-dynamic register modifications in the DDR self refresh
mode functions:
- hwlpctl.hw_lp_en & dfimisc.dfi_init_complete_en : group 3
- pwrtmg.selfref_to_x32 & powerdown_to_x32 : group 4

Adding sw_ack management is enough for group 4.
Use dedicated functions for Group 3.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 15ecdf265e55755aafb99ae0adae0cf5b5d6bfa6 fix(st-ddr): manage quasi-dynamic register modifications (stm32mp2)
a7b1aa25526e643c8d9a62947bf2679720498273 fix(st-ddr): manage quasi-dynamic register modifications]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I61d1b9be7b358be4722cac609267f48c70046fc0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/241831
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/ddr/stm32mp1_ddr.c         | 19 ++++++++++-
 drivers/st/ddr/stm32mp1_ddr_helpers.c | 46 ++++++++++++++++++++++++---
 2 files changed, 60 insertions(+), 5 deletions(-)

diff --git a/drivers/st/ddr/stm32mp1_ddr.c b/drivers/st/ddr/stm32mp1_ddr.c
index 1259ddf0c5..8b555a9386 100644
--- a/drivers/st/ddr/stm32mp1_ddr.c
+++ b/drivers/st/ddr/stm32mp1_ddr.c
@@ -528,21 +528,36 @@ static void stm32mp1_ddr3_dll_off(struct stm32mp_ddr_priv *priv)
 
 static void stm32mp1_refresh_disable(struct stm32mp_ddrctl *ctl)
 {
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * dfimisc.dfi_init_complete_en : Group 3
+	 * AXI ports not yet enabled, don't disable them
+	 */
+	stm32mp_ddr_disable_host_interface(ctl);
 	stm32mp_ddr_start_sw_done(ctl);
-	/* Quasi-dynamic register update*/
+
 	mmio_setbits_32((uintptr_t)&ctl->rfshctl3,
 			DDRCTRL_RFSHCTL3_DIS_AUTO_REFRESH);
 	mmio_clrbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_POWERDOWN_EN |
 						 DDRCTRL_PWRCTL_SELFREF_EN);
 	mmio_clrbits_32((uintptr_t)&ctl->dfimisc,
 			DDRCTRL_DFIMISC_DFI_INIT_COMPLETE_EN);
+
 	stm32mp_ddr_wait_sw_done_ack(ctl);
+	stm32mp_ddr_enable_host_interface(ctl);
 }
 
 static void stm32mp1_refresh_restore(struct stm32mp_ddrctl *ctl,
 				     uint32_t rfshctl3, uint32_t pwrctl)
 {
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * dfimisc.dfi_init_complete_en : Group 3
+	 * AXI ports not yet enabled, don't disable them
+	 */
+	stm32mp_ddr_disable_host_interface(ctl);
 	stm32mp_ddr_start_sw_done(ctl);
+
 	if ((rfshctl3 & DDRCTRL_RFSHCTL3_DIS_AUTO_REFRESH) == 0U) {
 		mmio_clrbits_32((uintptr_t)&ctl->rfshctl3,
 				DDRCTRL_RFSHCTL3_DIS_AUTO_REFRESH);
@@ -557,7 +572,9 @@ static void stm32mp1_refresh_restore(struct stm32mp_ddrctl *ctl,
 	}
 	mmio_setbits_32((uintptr_t)&ctl->dfimisc,
 			DDRCTRL_DFIMISC_DFI_INIT_COMPLETE_EN);
+
 	stm32mp_ddr_wait_sw_done_ack(ctl);
+	stm32mp_ddr_enable_host_interface(ctl);
 }
 
 static void stm32mp1_refresh_cmd(struct stm32mp_ddrctl *ctl)
diff --git a/drivers/st/ddr/stm32mp1_ddr_helpers.c b/drivers/st/ddr/stm32mp1_ddr_helpers.c
index 9bf5c1c8cf..f638452cd5 100644
--- a/drivers/st/ddr/stm32mp1_ddr_helpers.c
+++ b/drivers/st/ddr/stm32mp1_ddr_helpers.c
@@ -165,13 +165,17 @@ int ddr_sw_self_refresh_exit(void)
 	/* Enable all clocks */
 	ddr_enable_clock();
 
-	stm32mp_ddr_start_sw_done((struct stm32mp_ddrctl *)ddrctrl_base);
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * dfimisc.dfi_init_complete_en : Group 3
+	 */
+	stm32mp_ddr_set_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* Mask dfi_init_complete_en */
 	mmio_clrbits_32(ddrctrl_base + DDRCTRL_DFIMISC,
 			DDRCTRL_DFIMISC_DFI_INIT_COMPLETE_EN);
 
-	stm32mp_ddr_wait_sw_done_ack((struct stm32mp_ddrctl *)ddrctrl_base);
+	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* Switch controller clocks (uMCTL2/PUBL) to DLL ref clock */
 	stm32mp1_clk_rcc_regs_lock();
@@ -230,13 +234,17 @@ int ddr_sw_self_refresh_exit(void)
 		}
 	}
 
-	stm32mp_ddr_start_sw_done((struct stm32mp_ddrctl *)ddrctrl_base);
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * dfimisc.dfi_init_complete_en : Group 3
+	 */
+	stm32mp_ddr_set_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* Unmask dfi_init_complete_en to uMCTL2 */
 	mmio_setbits_32(ddrctrl_base + DDRCTRL_DFIMISC,
 			DDRCTRL_DFIMISC_DFI_INIT_COMPLETE_EN);
 
-	stm32mp_ddr_wait_sw_done_ack((struct stm32mp_ddrctl *)ddrctrl_base);
+	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* Deactivate sw retention in PWR */
 	stm32mp_pwr_regs_lock();
@@ -356,6 +364,14 @@ static void ddr_sr_mode_ssr(void)
 
 	stm32mp1_clk_rcc_regs_unlock();
 
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * hwlpctl.hw_lp_en : Group 3
+	 * pwrtmg.selfref_to_x32 & powerdown_to_x32 : Group 4
+	 * Group 3 is the most restrictive, apply its conditions for all
+	 */
+	stm32mp_ddr_set_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/* Disable HW LP interface of uMCTL2 */
 	mmio_clrbits_32(ddrctrl_base + DDRCTRL_HWLPCTL,
 			DDRCTRL_HWLPCTL_HW_LP_EN);
@@ -365,6 +381,8 @@ static void ddr_sr_mode_ssr(void)
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_MASK,
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_0);
 
+	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/*
 	 * Disable Clock disable with LP modes
 	 * (used in RUN mode for LPDDR2 with specific timing).
@@ -399,6 +417,14 @@ static void ddr_sr_mode_asr(void)
 
 	stm32mp1_clk_rcc_regs_unlock();
 
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * hwlpctl.hw_lp_en : Group 3
+	 * pwrtmg.selfref_to_x32 & powerdown_to_x32 : Group 4
+	 * Group 3 is the most restrictive, apply its conditions for all
+	 */
+	stm32mp_ddr_set_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/* Enable HW LP interface of uMCTL2 */
 	mmio_setbits_32(ddrctrl_base + DDRCTRL_HWLPCTL,
 			DDRCTRL_HWLPCTL_HW_LP_EN);
@@ -408,6 +434,8 @@ static void ddr_sr_mode_asr(void)
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_MASK,
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_0);
 
+	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/*
 	 * Enable Clock disable with LP modes
 	 * (used in RUN mode for LPDDR2 with specific timing).
@@ -442,6 +470,14 @@ static void ddr_sr_mode_hsr(void)
 
 	stm32mp1_clk_rcc_regs_unlock();
 
+	/*
+	 * Manage quasi-dynamic registers modification
+	 * hwlpctl.hw_lp_en : Group 3
+	 * pwrtmg.selfref_to_x32 & powerdown_to_x32 : Group 4
+	 * Group 3 is the most restrictive, apply its conditions for all
+	 */
+	stm32mp_ddr_set_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/* Enable HW LP interface of uMCTL2 */
 	mmio_setbits_32(ddrctrl_base + DDRCTRL_HWLPCTL,
 			DDRCTRL_HWLPCTL_HW_LP_EN);
@@ -451,6 +487,8 @@ static void ddr_sr_mode_hsr(void)
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_MASK,
 			   DDRCTRL_PWRTMG_SELFREF_TO_X32_0);
 
+	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
+
 	/*
 	 * Enable Clock disable with LP modes
 	 * (used in RUN mode for LPDDR2 with specific timing).
-- 
2.53.0

